### PR TITLE
Fix race condition for network card configuration 

### PR DESCRIPTION
**Pull request by jcvenegas**
This commit adds a function called wait_for_nic(char *nic_name).
This function will check if a network card exist in the the virtual
machine, if not it will wait until the device is detected.

Use udev events will prevent race conditions trying to use
a network card that is not detected yet.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

### CODE REVIEW

**Comment by jcvenegas** [comment]
Fixes https://github.com/01org/cc-oci-runtime/issues/623

File: src/net.c [hunk 1]
```diff
@@ -174,6 +174,74 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+		}
+
+		while (index < len) {
+			index+= snprintf(b, sizeof(b),"%s", buf+index);
+			index++;
+			char *p = strstr(b, "DEVPATH=");
+			if (!p) { continue; }
+			printf("%s\n", p);
+			fprintf(stdout, "%s\n", p);
+			char *d = strstr(b, nic_name);
```
Comment (mcastelino) [inline src/net.c]: will this give a false positive on eth0 if you get eth01

File: src/net.c [hunk 2]
```diff
@@ -174,6 +174,74 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+	struct sockaddr_nl nls;
+	char buf[512];
+	int fd = -1;
+	int ret = -1;
+	char path[512];
+
+	sprintf(path, "/sys/class/net/%s/ifindex", nic_name);
+
+	if( access( path, F_OK ) != -1 ) {
+		fprintf(stdout, "test: nic %s exist\n", nic_name);
```
Comment (devimc) [inline src/net.c]: missing 'return' ?

File: src/net.c [hunk 3]
```diff
@@ -174,6 +174,74 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+	}
+
+	memset(&nls,0,sizeof(struct sockaddr_nl));
```
Comment (devimc) [inline src/net.c]: ```sizeof(nls)``` is better

File: src/net.c [hunk 4]
```diff
@@ -174,6 +174,74 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+	nls.nl_pid = getpid();
+	nls.nl_groups = -1;
+
+	fd = socket(PF_NETLINK, SOCK_DGRAM, NETLINK_KOBJECT_UEVENT);
+	if (fd==-1){
+		perror("Open socket failed\n");
+		return -1;
+	}
+
+	if( access( path, F_OK ) != -1 ) {
```
Comment (amshinde) [inline src/net.c]: I see two access checks. This one can be removed.

Comment (devimc) [inline src/net.c]: missing space between if and (  ?

File: src/net.c (hunk 2 above)
Comment (jcvenegas) [reply src/net.c]: Actually this was a test message, removed  ...

File: src/net.c [hunk 5]
```diff
@@ -174,6 +174,70 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+			goto out;
+		}
+
+		while (index < len) {
+			index+= snprintf(b, sizeof(b),"%s", buf+index);
+			index++;
+			char *p = strstr(b, "DEVPATH=");
+			if (!p) { continue; }
+			printf("%s\n", p);
+			fprintf(stdout, "%s\n", p);
```
Comment (devimc) [inline src/net.c]: double printf

File: src/net.c (hunk 4 above)
Comment (jcvenegas) [reply src/net.c]: Removed the other, I check after I open the PF_NETLINK socket to avoid miss events related with the nic.

Comment (amshinde) [reply src/net.c]: Looks like I was looking at an earlier commit. The access check can be moved before opening socket itself.

File: src/net.c [hunk 6]
```diff
@@ -174,6 +174,70 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+{
+	struct sockaddr_nl nls;
+	char buf[512];
+	int fd = -1;
+	int ret = -1;
+	char path[512];
+
+	sprintf(path, "/sys/class/net/%s/ifindex", nic_name);
+
+	memset(&nls,0,sizeof(struct sockaddr_nl));
```
Comment (devimc) [inline src/net.c]: ```sizeof(nls)``` looks better

File: src/net.c [hunk 7]
```diff
@@ -174,6 +174,70 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+		return -1;
+	}
+
+	if( access( path, F_OK ) != -1 ) {
+		ret = 0;
+		fprintf(stdout, "nic %s exist\n", nic_name);
+		goto out;
+	}
+
+	if (bind(fd, (void *)&nls, sizeof(struct sockaddr_nl))) {
```
Comment (devimc) [inline src/net.c]: ```sizeof(nls)``` looks better

File: src/net.c (hunk 4 above)
Comment (jcvenegas) [reply src/net.c]: fixed

File: src/net.c (hunk 5 above)
Comment (jcvenegas) [reply src/net.c]: fixed

File: src/net.c (hunk 1 above)
Comment (amshinde) [reply src/net.c]: @mcastelino This is the predictable nic-name according to pci slot before it is changed later to eth0..

Comment (jcvenegas) [reply src/net.c]: nice catch, let me rework

File: src/net.c (hunk 4 above)
Comment (amshinde) [reply src/net.c]: If you do not want to miss any events, then shouldnt the access check go after the bind call?

File: src/net.c (hunk 1 above)
Comment (mcastelino) [reply src/net.c]: @amshinde what about enp0s25 vs enp0s2.

Comment (amshinde) [reply src/net.c]: @mcastelino Yes I agree.

Comment (mcastelino) [reply src/net.c]: @jcvenegas can you trigger off of the macaddress here? that is pretty unique. I think udev does provide macaddress info also. ID_NET_NAME_MAC

File: src/net.c [hunk 8]
```diff
@@ -174,6 +174,69 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
 	return 0;
 }
 
+static int wait_for_nic(const char *nic_name)
+{
+	struct sockaddr_nl nls;
+	char buf[512];
+	int fd = -1;
+	int ret = -1;
+	char path[512];
```
Comment (jodh-intel) [inline src/net.c]: `PATH_MAX` from `limits.h` would be safer here.

File: src/net.c [hunk 9]
```diff
@@ -174,6 +174,69 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
 	return 0;
 }
 
+static int wait_for_nic(const char *nic_name)
```
Comment (jodh-intel) [inline src/net.c]: Missing test for `nic_name != NULL && nic_name[0]`?

File: src/net.c [hunk 10]
```diff
@@ -174,6 +174,69 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+{
+	struct sockaddr_nl nls;
+	char buf[512];
```
Comment (jodh-intel) [inline src/net.c]: `BUFSIZ`?

File: src/net.c [hunk 11]
```diff
@@ -174,6 +174,69 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+	if (bind(fd, (void *)&nls, sizeof(struct sockaddr_nl))) {
+		perror("Bind failed\n");
+		goto out;
+	}
+
+
+	while (1){
+		int  index = 0;
+		int  len   = 0;
+		char b[512];
```
Comment (jodh-intel) [inline src/net.c]: `BUFSIZ`

File: src/net.c [hunk 12]
```diff
@@ -174,6 +174,69 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+
+		len = recv(fd, buf, sizeof(buf), 0);
+		if (len == -1){
+			perror("fail to read from UEVENT netlink socket \n");
+			goto out;
+		}
+
+		while (index < len) {
+			index+= snprintf(b, sizeof(b),"%s", buf+index);
```
Comment (jodh-intel) [inline src/net.c]: Assumes `snprintf()` worked - it might return `-1`.

File: src/net.c [hunk 13]
```diff
@@ -174,6 +174,69 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+	nls.nl_pid = getpid();
+	nls.nl_groups = -1;
+
+	fd = socket(PF_NETLINK, SOCK_DGRAM, NETLINK_KOBJECT_UEVENT);
+	if (fd==-1){
+		perror("Open socket failed\n");
+		return -1;
+	}
+
+	if ( access( path, F_OK ) != -1 ) {
```
Comment (sameo) [inline src/net.c]: * `if (access(path, F_OK) != -1)` for consistency sake.
* You should call `access` even before creating the socket.

File: src/net.c [hunk 14]
```diff
@@ -174,6 +174,69 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+		ret = 0;
+		fprintf(stdout, "nic %s exist\n", nic_name);
```
Comment (sameo) [inline src/net.c]: `nic %s already exists`

File: src/net.c (hunk 6 above)
Comment (sameo) [reply src/net.c]: I personally prefer `sizeof(type)` when the variable is not an array.

File: src/net.c [hunk 15]
```diff
@@ -174,6 +174,69 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+{
+	struct sockaddr_nl nls;
+	char buf[512];
+	int fd = -1;
+	int ret = -1;
+	char path[512];
+
+	sprintf(path, "/sys/class/net/%s/ifindex", nic_name);
+
+	memset(&nls,0,sizeof(struct sockaddr_nl));
```
Comment (sameo) [inline src/net.c]: `memset(&nls, 0, sizeof(struct sockaddr_nl));`

File: src/net.c [hunk 16]
```diff
@@ -174,6 +174,69 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+			if (!p) { continue; }
+			fprintf(stdout, "%s\n", p);
+			char *d = strstr(b, nic_name);
+			if (d) {
+				fprintf(stdout, "nic %s detected\n", nic_name);
+				ret = 0;
+				goto out;
+			}
+		}
+	}
```
Comment (sameo) [inline src/net.c]: IMHO, a more elegant solution for parsing a uevent would look like that:

```C
struct uevent {
	char *action;
	char *modalias;
	char *driver;
	char *devpath;
}

void parse_uevent(const char *msg, struct uevent *uevent)
{
	while (*msg) {
		if (!uevent->action && !strncmp(msg, "ACTION=", 7)) {
			msg += 7;
			uevent->action = strdup(msg);	
		} else if (!uevent->devpath && !strncmp(msg, "DEVPATH=", 8)) {
			msg += 8;
			uevent->devpath = strdup(msg);
		} 

		/* advance to after the next \0 */
		while (*msg++)
			;
	}
}
```

File: src/net.c [hunk 17]
```diff
@@ -774,9 +841,6 @@ int hyper_setup_network(struct hyper_pod *pod)
 	struct hyper_route *rt;
 	struct rtnl_handle rth;
 
-	if (hyper_rescan() < 0)
-		return -1;
-
```
Comment (sameo) [inline src/net.c:779]: Can we completely get rid of the PCI bus explicit scan ?

File: src/net.c (hunk 1 above)
Comment (jcvenegas) [reply src/net.c]: @mcastelino , yes the ID_NET_NAME_MAC is provided in the udev event, currently the hyperstat does not about the nic address when it is configured, we could add the mac address to the networkcard json that is provided to hyperstart and look for it.

File: src/net.c (hunk 16 above)
Comment (jcvenegas) [reply src/net.c]: totally agree, fixing :)

File: src/net.c (hunk 13 above)
Comment (amshinde) [reply src/net.c]: @sameo I think access needs to be called at two places, once could be called before the socket creation. (to avoid the socket creation and binding) and the other place it definitely needs to be called is  after the bind call to make sure no udev events are missed till bind() is called.

Comment (sameo) [reply src/net.c]: @amshinde You're assuming the kernel will queue the uevents after you bound the AF_NETLINK socker ?

**Comment by sameo** [comment]
~~LGTM~~
@jodh found a few bugs in the code still.

File: src/net.c [hunk 18]
```diff
@@ -174,6 +176,130 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+        while (*msg) {
+                if (!ue->action && !strncmp(msg, ACTION, strlen(ACTION))) {
+                        msg += strlen(ACTION);
+                        ue->action = strdup(msg);
+                } else if (!ue->devpath && !strncmp(msg, DEVPATH, strlen(DEVPATH))) {
+                        msg += strlen(DEVPATH);
+                        ue->devpath = strdup(msg);
+                } else if (!ue->modalias && !strncmp(msg, MODALIAS, strlen(MODALIAS))) {
+                        msg += strlen(MODALIAS);
+                        ue->devpath = strdup(msg);
```
Comment (jodh-intel) [inline src/net.c]: This looks like a bug to me as `devpath` is covered above?

(Not a blocker, but given that this entire block is very formulaic, a better approach would probably be to use an array of structs and maybe a macro to condense the code further)l.

File: src/net.c [hunk 19]
```diff
@@ -14,6 +14,8 @@
 #include "parse.h"
 #include "../config.h"
 
+#define UEVENT_BUFFER_SIZE 512
+
 void hyper_set_be32(uint8_t *buf, uint32_t val)
 {
 	buf[0] = val >> 24;
```
Comment (jodh-intel) [inline src/net.c:21]: Check on `buf != NULL`?

File: src/net.c [hunk 20]
```diff
@@ -174,6 +176,130 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+
+static int free_uevent(struct uevent *ue){
+	if(! ue){
+		return -1;
+	}
+        if ( ue->action) { free(ue->action); ue->action = NULL; }
+        if ( ue->modalias) { free(ue->modalias); ue->action = NULL; }
+        if ( ue->driver) { free(ue->driver); ue->driver = NULL; }
+        if ( ue->devpath ) { free(ue->devpath); ue->devpath = NULL; }
+        if ( ue->interface) { free(ue->interface); ue->interface = NULL; }
```
Comment (jodh-intel) [inline src/net.c]: A macro would allow this code to be simplified.

File: src/net.c [hunk 21]
```diff
@@ -174,6 +176,130 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
 	return 0;
 }
 
+
+static int free_uevent(struct uevent *ue){
+	if(! ue){
+		return -1;
+	}
+        if ( ue->action) { free(ue->action); ue->action = NULL; }
+        if ( ue->modalias) { free(ue->modalias); ue->action = NULL; }
```
Comment (jodh-intel) [inline src/net.c]: Looks like a bug here too.

File: src/net.c (hunk 18 above)
Comment (sameo) [reply src/net.c]: Ah, good catch. That's a bug indeed.

File: src/net.c [hunk 22]
```diff
@@ -174,6 +176,130 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+static int wait_for_nic(const char *nic_name)
+{
+	struct sockaddr_nl nls;
+	char msg[UEVENT_BUFFER_SIZE];
+	int fd = -1;
+	int ret = -1;
+	char path[PATH_MAX];
+	struct uevent ue = {0};
+
+	if (nic_name != NULL && nic_name[0]) {
```
Comment (jodh-intel) [inline src/net.c]: `if (! (nic_name && nic_name[0])) { return -1; }` ?

File: src/net.c (hunk 20 above)
Comment (devimc) [reply src/net.c]: yeah, something like
```
#define free_if_set(e) if (e) { free(e); e=NULL; }
```

File: src/net.c [hunk 23]
```diff
@@ -174,6 +176,130 @@ static int addattr_l(struct nlmsghdr *n, int maxlen, int type, void *data, int a
...
+        if( !(msg && ue) ) {
+                return -1;
+        }
+        if( len < 0 ) {
+                return -1;
+        }
+
+        while (*msg) {
+                if (!ue->action && !strncmp(msg, ACTION, strlen(ACTION))) {
+                        msg += strlen(ACTION);
```
Comment (devimc) [inline src/net.c]: you can also use ```sizeof(ACTION)-1``` is safer but you have change ACTION declaration to ```const char ACTION[] = "ACTION=";```

File: src/net.c (hunk 1 above)
Comment (jcvenegas) [reply src/net.c]: @mcastelino changed the way to parse the udev event, now would not give the false positive

File: src/net.c (hunk 4 above)
Comment (jcvenegas) [reply src/net.c]: called after bind

File: src/net.c (hunk 8 above)
Comment (jcvenegas) [reply src/net.c]: fixed

File: src/net.c (hunk 10 above)
Comment (jcvenegas) [reply src/net.c]: fixed

File: src/net.c (hunk 14 above)
Comment (jcvenegas) [reply src/net.c]: fixed

File: src/net.c (hunk 18 above)
Comment (jcvenegas) [reply src/net.c]: nice catch !, @jodh-intel  changed for a loop +  array of structs

File: src/net.c (hunk 19 above)
Comment (jcvenegas) [reply src/net.c:21]: maybe can create new PR for those :)

File: src/net.c (hunk 20 above)
Comment (jcvenegas) [reply src/net.c]: @devimc agree :) added free_if_set to util.h

File: src/net.c (hunk 21 above)
Comment (jcvenegas) [reply src/net.c]: fixed

File: src/net.c (hunk 22 above)
Comment (jcvenegas) [reply src/net.c]: fixed

**Comment by sameo** [comment]
LGTM
